### PR TITLE
add configure_dynamic_scaling stage to edxapp pipeline

### PR DESCRIPTION
**Pull request by pwnage101**
WIP.

Add a stage which runs the `configure_dynamic_scaling.py` tubular script which deploys cloudwatch alarms and scaling policies causing the edxapp ASG to dynamically scale.

This ties in the following tubular PR: https://github.com/edx/tubular/pull/182

### CODE REVIEW

**Comment by MichaelRoytman** [comment]
@edx/pipeline-team Could you review this PR to make sure this will accomplish adding a "configure_dynamic_scaling" stage to the edxapp pipeline? The relevant ticket is here: https://openedx.atlassian.net/browse/DEVOPS-6177.

File: edxpipelines/patterns/stages.py [hunk 1]
```diff
@@ -418,6 +418,63 @@ def generate_deploy_ami(
...
+    tasks.generate_package_install(job, 'tubular')
+    # Make the artifact directory if it does not exist
+    job.add_task(ExecTask(
+        [
+            '/bin/bash',
+            '-c',
+            'mkdir -p ../{}'.format(constants.ARTIFACT_PATH),
+        ],
+        working_dir="tubular"
+    ))
```
Comment (doctoryes) [inline edxpipelines/patterns/stages.py]: Please use `tasks.generate_target_directory()` here instead.

File: edxpipelines/patterns/stages.py [hunk 2]
```diff
@@ -418,6 +418,63 @@ def generate_deploy_ami(
...
+        constants.CONFIGURE_DYNAMIC_SCALING_OUT_FILENAME,
+    )
+    job.ensure_artifacts(set([BuildArtifact(artifact_path)]))
+
+    deploy_command = 'configure_dynamic_scaling.py '
+    deploy_command += '--out_file ../{} '.format(artifact_path)
+    deploy_command += '--environment {} '.format(edx_environment)
+
+    # Execute the deployment script
+    job.add_task(ExecTask(['/bin/bash', '-c', deploy_command], working_dir="tubular"))
```
Comment (doctoryes) [inline edxpipelines/patterns/stages.py]: Please use `tasks.tubular_task()` here - like this example:
https://github.com/edx/edx-gomatic/blob/master/edxpipelines/patterns/stages.py#L1161-L1171

File: edxpipelines/pipelines/cd_edxapp.py [hunk 3]
```diff
@@ -201,6 +201,18 @@ def install_pipelines(configurator, config):
...
 
+    #
+    # Create the stage to configure the ASG dynamic scaling.
+    #
+    stages.generate_configure_dynamic_scaling(
+        pipeline,
+        config['aws_access_key_id'],
+        config['aws_secret_access_key'],
+        config['edx_environment'],
+        # TODO: create a new config variable for enabling manual approval for this stage.
```
Comment (doctoryes) [inline edxpipelines/pipelines/cd_edxapp.py]: Is there a reason not to create a new config var now?

Comment (pwnage101) [inline edxpipelines/pipelines/cd_edxapp.py]: No reason other than the fact that this is the first time I ever wrote pipeline code and I didn't want to invest effort down a potentially bad path.  So far you don't seem to have any major complaints, so I'll just address this TODO now.

Comment (pwnage101) [inline edxpipelines/pipelines/cd_edxapp.py]: I'm actually confused about the `auto_*` config variables.  IIRC, when i click to start the loadtest pipeline, it doesn't wait for my approval before deploying the AMI, but in `gocd/vars/tools/environment-deployment-play/loadtest-edx-edxapp.yml` in the edx/gocd-secure repo this line is present:

```
auto_deploy_ami: "False"
```

Which suggests that I should have been consulted for approving the stage.  Is there something I'm missing?

Comment (doctoryes) [inline edxpipelines/pipelines/cd_edxapp.py]: So - that behavior is due to a bug that this change will fix:
https://github.com/edx-ops/gomatic-secure/pull/130

To find this bug, here's the steps I used:
- Look at the config.yml file to find the section that builds the pipeline.
- Duplicate the call to build the pipeline on the command-line as below. Use the "--dry-run" option:
`python edxpipelines/pipelines/cd_edxapp.py --dry-run --variable_file ../gomatic-secure/gocd/vars/tools/admin.yml --variable_file ../gomatic-secure/gocd/vars/tools/environment-deployment-play/loadtest-edx-edxapp.yml --variable_file ../gomatic-secure/gocd/vars/tools/environment-deployments/loadtest-edx.yml --variable_file ../gomatic-secure/gocd/vars/tools/environment-play/loadtest-edxapp.yml --variable_file edxpipelines/pipelines/config/edxapp.yml --variable_file ../gomatic-secure/gocd/vars/tools/deployments/edx.yml`
- Connect to the VPN.
- Add any needed breakpoints to the gomatic code, like:
`import pdb;pdb.set_trace()`
- Run the command (in a virtualenv with all requirements installed).
- Examine things in the debugger.

**Comment by pwnage101** [comment]
@doctoryes ready for review again.  BTW, I will not merge until after testing and updating the commit message.

File: edxpipelines/pipelines/cd_edxapp.py [hunk 4]
```diff
@@ -201,6 +201,17 @@ def install_pipelines(configurator, config):
...
 
+    #
+    # Create the stage to configure the ASG dynamic scaling.
+    #
+    stages.generate_configure_dynamic_scaling(
+        pipeline,
+        config['aws_access_key_id'],
+        config['aws_secret_access_key'],
+        config['edx_environment'],
+        manual_approval=not config.get('auto_configure_dynamic_scaling', False)
```
Comment (doctoryes) [inline edxpipelines/pipelines/cd_edxapp.py]: Remember to make this a true YAML boolean in the config - not a string literal.

**Comment by pwnage101** [comment]
Closing for the same reason as https://github.com/edx/tubular/pull/182

I'll leave this branch lingering for future reference.

**Pull request closed without merging**